Data management layer for MSPileup data-service

vkuznet opened the pull request:
Fixes #11424 

#### Status
In development

#### Description
New `MSPileupData.py` module provides data management layer for `MSPileup` data service. In particular, it implements the following set of APIs to deal with MSPileup data:
- `def create(self, cdict)` creates MSPileup data from given campaign.json dictionary, and stores it to persistent storage
- `def update(self, data)` updates MSPileup data in persistent storage with given data dictionary
- `def store(self, data)` stores MSPileup data to persistent storage, e.g. someone can provide pileup data to store
- `def get(self, pileupId)` fetches MSPileup data from persistent storage for a given pileupID
- `def delete(self, pileupId)` deletes MSPileup data in persistent storage for given pileupID
- `def validate(self, pdict)` to validate input data dictionary

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs

#### External dependencies / deployment changes

vkuznet committed:
MSPileup service data management layer

vkuznet committed:
Unit test for MSPileupData layer

vkuznet commented:
@amaltaro , @todor-ivanov I put in this PR initial implementation for data management layer. This is work in progress as I need your initial feedback on the following questions:
- which database the data should be stored to, so far I put MongoDB as backend database
- how data should be created, in #11424 there is an initial data-structure but it remains unclear how data should be created. I already looked at campaign.json and it remains unclear how it should be used. Please provide concrete example. For instance, we may create a dictionary out of proposed data structure, but I need input who will fill out the necessary values. Or, the data will be given from upstream caller.

Once, we'll clarify these I can proceed to adjusting the codebase. Meanwhile, I provided an initial unit test for the data management class.

amaltaro commented:
@vkuznet I provided further context and information in the GH issue itself. Please have a look.

todor-ivanov commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,177 @@
+from WMCore.MicroService.MSCore.MSCore import MSCore
+
+
+def stripKeys(docs, stripKeys=['_id']):
```
I suppose this logic here is needed in order to strip the `_id` field from the Mongo documents. This has been already addressed in other way in the rest of the microservices using Mongo DB by explicitely defining a document schema in which the `_id` is present, hence it is preserved in both Mongo docmuemnt  and the python interpretation of the document. We better think of sharing logic.

vkuznet replied:
So far, stripKeys exist in other MS services but those are internal service API. I agree it is common function and most likely it should go into MongoDB module and called from there. I'll be happy to put it into MongoDB module though if you agree.

todor-ivanov replied:
Yes indeed, not a bad idea for this piece to go into the MongoDB module wrapper itself. 

I can admit, in the past it was done in at least two different ways, for two different modules. A quick lookup can lead us to: 
https://github.com/dmwm/WMCore/blob/479146e4ae1a73a91f44b9ab0ea8fe36195a2dae/src/python/WMCore/MicroService/MSOutput/MSOutput.py#L959-L968 
and 
https://github.com/dmwm/WMCore/blob/479146e4ae1a73a91f44b9ab0ea8fe36195a2dae/src/python/WMCore/MicroService/MSUnmerged/MSUnmergedRSE.py#L72-L90

They are addressing some slightly different use cases. And I am more in favor of  the later, since it is more friendly to the model where we have a strictly defined docSchema. And the projection can be put directly into the MongoDB query. But this one was also covering some subtle details in MSUnmerged, so we need to be careful if we decide to go for unification of this method.

todor-ivanov commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,177 @@
+        mongoDB = MongoDB(**msOutDBConfig)
+        self.msDB = getattr(mongoDB, self.msConfig['mongoDB'])
+        self.dbColl = self.msDB[collection]
+#         self.dbColl.create_indexes([mongoIndex])
```
Following on what I've already written in here:https://github.com/dmwm/WMCore/issues/11424#issuecomment-1381935142
Everything in this `_init_` up to this point, I'd say needs to go in the service  and a single database session gets maintained per service poling cycle so we reuse resources. If we want to parallelize the queries to the database  there are other ways to do it, but not to create a single session with every document/object - if we go with a single document per PU this will increase a lot both memory footprint  and network usage.

vkuznet replied:
Todor, I disagree with your observation. Here is how I see separation of service and database layers with single session:
```
class MSPileupService():
    def __init__(self):
           # it keeps internal access to DB but it is re-used in this class across APIs, i.e. single session
           self.mspileup = MSPileupData()
    def postAPI(self, input):
          # it can call self.mspileup object to perform any database related operation, e.g.
          # create and store object
          cdict = validate(input)
          doc self.mspileup.create(cdict)
          self.mspileup.store(doc)
```
This way, you keep database details inside MSPileupData class and keep HTTP logic in Service class. There is no need to keep database logic within Service class. Each class responsible for its own behavior, i.e. Service class handles HTTP logic, while Data class handle database details.

todor-ivanov replied:
Well that's a matter of separation of code modules. All fine with me. I do not object having a separate file/module dealing with the database layer, as long as we have a single instance of it during the service execution and we  reuse the same object many times. I am actually more in favor of that -  this is  a sign of well organized code. Me likes! :)  

Then what I can sufggest to happen in this case is to create a third module, describing only the document schema and validation data logic, as done in the other micro services.  And this we cannot avoid, simply because there is a difference in the internal data interpretation between MongoDB and python - for this third one as described bellow we should have a single instance per pileup/object returned from the database.

vkuznet replied:
ok, then, we are in agreement.

todor-ivanov commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,177 @@
+
+    def create(self, cdict):
+        """
+        Create and return pileup data from campaigns dictionary
```
We'd better think of continuous parsing of the campaign config and updating of all relevant documents in the database based on the information obtained in the process and reflecting also the current  load in the system. So again, this needs to go as a step in the service.

vkuznet replied:
No, service can accept campaigns.json, parse it and only call data layer to insert docs, e.g.
```
class MSPileupService():
    def __init__(self):
           # it keeps internal access to DB but it is re-used in this class across APIs, i.e. single session
           self.mspileup = MSPileupData()
     def campaignAPI(self, campaigns):
          # parse campaigns:
          for doc in parse(campaigns:)
               doc = self.mspileup.createPileupObj(doc)
               self.mspileup.strorePileupObj(doc)
```
Nice and clear separation of layers, the service accept campaigns, and parse it. Then it calls data layer to insert docs.

todor-ivanov replied:
One note here, though. I wouldn't  call this method  `campaignAPI`, because it should not be supposed to provide inter-service communication API (I can imagine only internal service calls for this one). Maybe calling it  `campaingParser` would be more direct and pointing to it's exact and sole purpose. Just an idea...

todor-ivanov commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,177 @@
+        spec = {'pileupId': pileupId}
+        self.dbColl.remove(spec)
+
+    def validate(self, pdict):
```
this document validation also may get a similar approach to the way how we define those data structures in other micro services. Meaning if we have  a doc schema defined in the `_init_`  we could simply iterate  through it to check data types. similar to: 
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/MicroService/MSRuleCleaner/MSRuleCleanerWflow.py#L134-L230

todor-ivanov reviewed (changes requested):
Thanks  @vkuznet for this PR

Please take into account that all my comments in this review are made in the context of one of my previous comments  made here:
https://github.com/dmwm/WMCore/issues/11424#issuecomment-1381935142

So in general my opinion is that this module defined here needs to reflect only:
* document schema 
   +
* data validation logic
   + 
* (maybe) functionalities needed to parse different sources of information, like workflow descriptions and campaign configuration. 

All the rest needs to go in the service logic, including database sessions maintainenance and execution parallelization.

amaltaro commented:
Valentin, I fully second what Todor said here. Ideally, this data structure object (module) should not depend on MongoDB itself. Please follow the same structure as done for other microservices (WMCore/MicroService/MSPileup/DataStructs/yourmodule).

Actions like get/store/update/etc do belong to the main logic of MSPileup though.
Last but not least, please have a look at the original description of this issue, it has been updated after further discussion with the P&R. Nothing has been changed or removed, only new stuff added.

amaltaro reviewed (commented):
(Just getting rid of the review request, given that I provided input in non-review mode)

vkuznet commented:
> Valentin, I fully second what Todor said here. Ideally, this data structure object (module) should not depend on MongoDB itself. Please follow the same structure as done for other microservices (WMCore/MicroService/MSPileup/DataStructs/yourmodule).
> 
> Actions like get/store/update/etc do belong to the main logic of MSPileup though. Last but not least, please have a look at the original description of this issue, it has been updated after further discussion with the P&R. Nothing has been changed or removed, only new stuff added.

Alan, it is totally fine to put Data structure into MSPileup/DataStrucs as you proposed, and I agree to do that. And, I think we (you and I) are also agree that apis like `get/store/update` belong to data layer which draft is here. But I disagree with Todor's suggestion to put data layer logic (like MongoDB) into service since in my experience it is not right design architecture. The service should be responsible for HTTP logic and nothing else, it should call data layer to handle database actions via its APIs. And, database layer should handle database connection and interaction. If we need to scale then service will either setup data layer pool or ask data layer to use the pool of connection. Again, pool belongs to data lay

vkuznet commented:
I'm fine to put data structure into MSPileup/DataStructs, and I agree with Alan that `get/update/create/delete` APIs belong here.

todor-ivanov commented:
thanks @vkuznet and @amaltaro. Let me use the current issue to make a more general comment.

Well, I do not want to say it now, since we are fully booked with other issues to work on. But we do tend to go in a well defined path with all services using MongoDB as a background database for keeping service states between runs. 

We basically define a strict document schema for every service and we can do bidirectional validation for inbound and outbound  database communication. 

This does lead us in a situation with few pros and cons: 
* We do have a simple and reliable model for using the Database 
* We do have a very good database integrity
* We do provide a base for code and methods unification
* We unfortunately, somewhat, restrain ourselves from using complex return result aggregations and utilize in full steam all the benefits provided by the database of choice.   

While for the last of those bullets, we can think of optimizations in near or long future (even though for the purposes we use the database it is not mandatory), for the third one we  can take actions immediately. We may think of generalizing some (if not all) of those document validation methods based on the document schema we  provide (as defined in the micro service data structure modules - e.g `MSRuleCleanerRSE, `MSOutpuTemplate`, etc...) and put those methods directly in the `WMCore.Database.MongoDB` wrapper module. This way, we can achieve even more flexibility for future changes of data structure schema for any micro service, which would be transformed into changes only in the relevant `DocSchema`, while any database peculiarities and service specific methods should be left, as @vkuznet suggests, in a separate database management layer for the particular micro service (as here).

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , @todor-ivanov please review a new edition of this PR. It includes full implementation of APIs based on our previous discussion. In particular:
- `def create(self, pdict)`
- `def update(self, doc)`
- `def get(self, spec)`
- `def delete(self, spec)`

Please note, that I opted out to use `spec` (JSON query) in `get and delete` APIs. Said that, initially I'll only implement `spec={'pileupName': <value>}` queries in HTTP layer which will pass this spec to data layer API. This will allow to maintain flexible queries via spec and at the same time constrain in HTTP layer the queries from end-users to one we discussed initially. When, we'll be ready to support more queries this layer will already be ready for that.

I also added time  serialization functions to support RFC339 for presentation layer. Again, this serialization will be used in HTTP layer of APIs to present time in human readable form, while at data layer everything is processed in seconds since epoch in GMT time zone.

I provided unit tests and ensure that code scores 10.00 in pylint.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+        try:
+            obj = MSPileupObj(pdict)
+            doc = obj.data
+        except Exception as exp:  # pylint: disable=broad-except
```
please don't disable pylint in the middle of the module. Actually, I would insist that we should not disable pylint messages.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+            obj = MSPileupObj(pdict)
+            doc = obj.data
+        except Exception as exp:  # pylint: disable=broad-except
+            msg = f"Failed to create MSPileupObj, {exp}"
```
we might have to cast `exp` to a string.

amaltaro replied:
I could be wrong, but this error message would print only the exception name, not the content/message that comes with it. Hence the suggestion above. But you might want to check it before making any changes.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+        pname = doc.get('pileupName', '')
+        if not pname:
+            err = MSPileupError(doc, f'No document found for {pname}')
+            return [err.NoKeyFound()]
```
I know we discussed this on Monday and it was the agreement for propagating an error to the end user. However, if we implement it at this layer, it will increase the complexity of the MSPileup daemon, because the same list object returned by this method can either contain an error object, or the pileup object itself.

I wonder if this logic should be implemented in the layer above, aka the MSPileup daemon/app running the whole logic of the service. For instance:
1. read data from the database
2. execute the logic of data placement and check
3. update the pileup object
4. persist the pileup object in the database

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+        # check if given document contains pileup Name (unique key)
+        pname = doc.get('pileupName', '')
+        if not pname:
+            err = MSPileupError(doc, f'No document found for {pname}')
```
You probably meant something like "No pileup name provided"

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+    return doc
+
+
+class MSPileupData(MSCore):
```
For this class, I would highly recommend to use more meaningful (and less confusing) method names, e.g.:
* get --> getPileup
* create --> createPileup
* and so on.

amaltaro replied:
Second commit. I see not reason why it should inherit from MSCore. This is only dealing with MongoDB, nothing else and it's not going to be the MSPileup daemon/app, AFAIU.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+        :return: list of MSPileupError or empty list
+        """
+        try:
+            self.dbColl.delete_one(spec)
```
How about adding a `self.logger.info("Pileup object %s successfully deleted from the database", spec.get("pileupName"))`?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+        # we do not need to create MSPileupObj and validate it since our doc comes from DB directly
+        doc['lastUpdateTime'] = gmtime()
+        try:
+            self.dbColl.update_one(spec, {"$set": doc})
```
How about adding a `self.logger.info("Pileup object %s successfully updated in the database", spec.get("pileupName"))`?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,204 @@
+
+        # insert document into underlying DB
+        try:
+            self.dbColl.insert_one(doc)
```
How about adding a `self.logger.info("Pileup object %s successfully created in the database", doc.get("pileupName"))`?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupError.py:
```diff
@@ -0,0 +1,110 @@
+        self.assign(msg=msg, code=MSPILEUP_DATABASE_ERROR)
+        return self.error()
+
+    def UniqueConstrainError(self):
```
Even though these methods are supposed to be an specific exception, I wonder whether we should stick to the .pylintrc convention and start their names with a lower case char. What do you think?

vkuznet replied:
Alan, here and in general I follow python style convention. Said that, in python all exceptions are capital case, in other 3rd party modules, e.g. pymongo, they are in capital case, etc. Therefore, I think we should be consistent with python style. If you still insist on making them lower case please say so and I'll change that, I just wanted to reply to your comment with reason I had in mind and let you think about it one more time.

vkuznet replied:
done, new code contains child exception classes

amaltaro reviewed (changes requested):
Valentin, this implementation is taking shape and looking good. However, I left many comments, suggestions and questions along the code.

In addition to those:
1) I understand MSPileupData is only handling with the database layer. In other words, it's not going to be the daemon/app that will be executing the full logic of the service. In that case, please remove the inheritance on MSCore.
2) Please review the data structure - now fully updated - in the GH issue. There are a few things missing in the expected schema.
3) there are quite a few typos here and there (docstrings and comments). Please review them before I start making 10s of comments to point them out.

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileup.py:
```diff
@@ -0,0 +1,134 @@
+"""
+File       : MSPileup.py
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileup.py:
```diff
@@ -0,0 +1,134 @@
+    :return: a list of tuples
+    """
+    doc = [
+        ('pileupName', None, (bytes, str)),
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileup.py:
```diff
@@ -0,0 +1,134 @@
+from WMCore.MicroService.Tools.Common import getMSLogger
+
+
+def schema():
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileup.py:
```diff
@@ -0,0 +1,134 @@
+    "active": boolean,
+    "pileupSizeGB": integer, current size of the pileup in GB (integer)
+    "rulesList: list of strings (rules) used to lock the pileup id
+    "campaigns": [ {"name": CampaignStingName , "active": boolean}, ... ]
```
done

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , I addressed all issues you requested and this PR is ready for review. Please note, I see that unit tests are failing but those seems to be unstable tests and not realted to this PR. The score of the code is decreased due to some of the features you requested, e.g. put schema/validation back to class instead of making them stand-along functions, etc. But I tried to best of my ability to fix all outstanding pylint/pep issues.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+            'lastUpdateTime': self.lastUpdateTime,
+            'expectedRSEs': self.expectedRSEs,
+            'fullReplicas': self.fullReplicas,
+            'campaigns': self.campaigns,
```
You named `campaignList` as `campaigns` here. Which is also fine for me, we just need to make it consistent everywhere (including the GH issue). Please take one and update things accordingly.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+            'activatedOn': self.activatedOn,
+            'deactivatedOn': self.deactivatedOn,
+            'active': self.active,
+            'pileupSize': self.pileupSize,
```
I see you added the two optional parameters. This was a good call, I really think they will be useful as it gets more mature.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+            'pileupType': self.pileupType,
+            'insertTime': self.insertTime,
+            'lastUpdateTime': self.lastUpdateTime,
+            'expectedRSEs': self.expectedRSEs,
```
This data structure is missing the `currentRSEs` parameter.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+from WMCore.MicroService.Tools.Common import getMSLogger
+
+
+def gmtimeSeconds():
```
Apologies for not mentioning this before, but the best place to have such function would be in this Utils module:
https://github.com/dmwm/WMCore/blob/master/src/python/Utils/Timers.py

Please make sure there is a relevant unit test for it.

vkuznet replied:
I'm not sure what should be relevant unit test in this case though since we only get current seconds in GMT time zone from default python library call.

amaltaro replied:
Hmm, indeed. Maybe something very basic like:
```
t1 = gmtimeSeconds()
time.sleep(1)
t2 = gmtimeSeconds()
self.assertTrue(isinstance(t2, int))
self.assertTrue(t2 > t1)
```
?

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        return json.dumps(self.data, indent=4)
+
+    def validate(self, pdict):
```
Valentin, we should implement value specific validation as well. A non-exhaustive list would be:
1) validate `pileupName` value to pass this Lexicon `dataset` method: https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/Lexicon.py#L208
2) `pileupType` in `('classic', 'premix')`
3) `expectedRSEs` and `currentRSEs`. My suggestion here is to make an specific method for this validation. My idea is that this `MSPileupObj` class accepts a list of RSEs as input parameter and uses it to validate RSE values (this way we do not need to add Rucio service dependency into a data structure module)
4) `containerFraction` between 0.0 and 1.0
5) `replicationGrouping` in `('DATASET', 'ALL')`

Please create the relevant unit tests.

vkuznet replied:
added initial implementation but need to provide unit tests for all values (will come back to it once I'll get time next week).

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        docSchema = self.schema()
+        for key, val in pdict.items():
+            msg = f"Failed to validate: {key}, expect data-type string got type=" + str(type(val))
```
I had a previous comment on this function pointing to two issues:
1) this error msg is wrong. Not all attributes are meant to be of string type
2) and how to make the complexity of this code cheaper (O(n) instead of O(n^2))

Please check your notifications and consider those as well. I can no longer see them when reviewing this PR, but I know I commented on it.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,215 @@
+        # first, create MSPileupObj which will be validated against its schema
+        try:
+            obj = MSPileupObj(pdict)
+            doc = obj.data
```
Please make a getter method to return the `data` attribute, instead of accessing it directly in the object.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,215 @@
+    return docs
+
+
+def encodeTimestamps(doc):
```
If we want to use the very same string format in different places of the project - which I think it would be great - then I would suggest to decompose this method and move the specific int - str time conversion under the Utils module as well:
https://github.com/dmwm/WMCore/blob/master/src/python/Utils/Timers.py

then in this function you keep iterating over those keys of the doc, but call the conversion function from the Timers module. In the Timers module, maybe call the function as `encodeGMTTimestamp`, just a suggestion thuogh.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/MSPileupData.py:
```diff
@@ -0,0 +1,215 @@
+    return doc
+
+
+def decodeTimestamps(doc):
```
Same suggestion as made above for the `encodeTimestamps` function.

vkuznet replied:
done

amaltaro reviewed (changes requested):
Thank you for providing all those changes, Valentin.
I left a few more comments along the code which involve a) must do changes and b) things that would make the overall code better and more reusable. Please have a look.

vkuznet commented:
Alan, thanks for review, I implemented some of the changes, but will proceed with further development next week when time permits.

vkuznet commented:
test this please

vkuznet commented:
@amaltaro , as far as I can tell all changes are in place and even pyling-py3k are not perfect I doubt nothing I can further improve here. The unit test are unstable and I always get different set of failures not related to this PR. Please review and let me know if any other changes will be required.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+                msg = f"replicationGrouping value {val} is neither of ['DATASET', 'ALL']"
+                self.logger.error(msg)
+                return False
+            if key == 'containerFraction' and isinstance(val, float) and (val > 1 or val < 0):
```
Isn't data type already validated in the beginning of this for loop?

vkuznet replied:
it is, so I removed `isinstance` call then.

amaltaro reviewed (changes requested):
@vkuznet please find further review along the code.

vkuznet commented on src/python/Utils/Timers.py:
```diff
@@ -7,9 +7,49 @@
+
+def encodeTimestamps(doc, keys):
+    """
+    Encode timestamps in MSPileup data document
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+            validRSEs = []
+        self.validRSEs = validRSEs
+
+        self.pileupName = pdict.get('pileupName', '')
```
ok, done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+        """
+        Data validator according to its schema
+
+        :param pdict: input data dictionary
```
Here I had in mind that we may use to validate external object, e.g. code creates MSPileupObj and then change something in dictionary and can use validate method to validate such dictionary without creating another object. I adjusted code to handle both cases.

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+                msg = f"pileupName value {val} does not match dataset pattern"
+                self.logger.error(msg)
+                return False
+            if key == "pileupType" and val not in ['classic', 'prefix']:
```
done

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+                msg = f"Failed to validate: {key}, expect data-type {stype} got type {dtype}"
+                self.logger.error(msg)
+                return False
+            if key == 'pileupName' and not dataset(val):
```
done, added try/except to check for this assertion error

vkuznet commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,174 @@
+               'currentRSEs': ([], list),
+               'fullReplicas': (0, int),
+               'campaigns': ([], list),
+               'containerFraction': (0.0, float),
```
done

vkuznet commented:
Alan, thanks for feedback, we're getting closer to a final version. To simplify your review task here are the changes I addressed based on your recent feedback:
- timers code changes can be found in this https://github.com/dmwm/WMCore/pull/11438/commits/edf90c0d94523c87cb1ae1655e9e91c5762e2f6f
- unit tests changes for timers code are here https://github.com/dmwm/WMCore/pull/11438/commits/03c69b5d041ea2e5870749f6ae28d5aeab1c37aa
- and, changes to MSPileupObj are here https://github.com/dmwm/WMCore/pull/11438/commits/3f6d9dd1d8cf2bba6f405cfe0eb718ceb431de65

Now, I'll squash them for the final set.

vkuznet commented:
Alan, I checked pylint/pep/unit tests (which are unstable) and up to my knowledge everything is ready. Please review again, and I posted individual commits for your convenience which addressed your recent review.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,165 @@
+
+        self.data = {
+            'pileupName': pdict.get('pileupName', ''),
+            'pileupType': pdict.get('pileupType', 'classic'),
```
I think the pileupType default value should be `""`. Thus ensuring that the end-user provides the appropriate configuration.

vkuznet replied:
done

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,165 @@
+            'currentRSEs': pdict.get('currentRSEs', []),
+            'fullReplicas': pdict.get('fullReplicas', 0),
+            'campaigns': pdict.get('campaigns', []),
+            'containerFraction': pdict.get('containerFraction', 0.0),
```
containerFraction default value should be `1.0`

vkuznet replied:
ahh, I fixed it in schema but not as default here. Thanks for spotting this again.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,165 @@
+            'activatedOn': pdict.get('activatedOn', gmtimeSeconds()),
+            'deactivatedOn': pdict.get('deactivatedOn', gmtimeSeconds()),
+            'active': pdict.get('active', False),
+            'pileupSize': pdict.get('pileupSize', 0),
```
I had a previous review on this `pileupSize` field. Please check your inbox/notifications.

vkuznet replied:
I updated https://github.com/dmwm/WMCore/issues/11424 that pileupSize should be in bytes. We should have additional function (if required) which can provide conversion from bytes to any other metric, like GB, but it is a decision which should be made outside the scope of this layer.

amaltaro commented on src/python/WMCore/MicroService/MSPileup/DataStructs/MSPileupObj.py:
```diff
@@ -0,0 +1,165 @@
+        if not pdict:
+            pdict = self.data
+        docSchema = self.schema()
+        for key, val in pdict.items():
```
Am I correct in saying that if the end user provides extra key/value pairs, that this validate function would not catch them and would let the request proceed? If so, then I think we should ensure that the end-users are providing the correct data structure, not more, not less fields.

vkuznet replied:
If user will provide a key:value pair which is not in a schema it will be caught in next line (https://github.com/dmwm/WMCore/pull/11438/files/afa5aa802ee6d41b4110d7d2fe3b8557830adbd0#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR92) and dict will not be validated.

vkuznet replied:
well, it is tricky to answer. Currently, if we create object and a key is not provided, then object will have default value defined in a schema for that key. But if we validate dictionary which is not created as object, then your observation may be valid. To avoid that I'll add the check for keys.

vkuznet replied:
the changes are committed.

amaltaro replied:
Valentin, you are right. If we always validate `self.data`, then the parameters will always match what is expected in the schema.

Thinking about a use POSTing a new pileup configuration, I think it would be ideal to validate the user-provided parameters, such that they know in case something is missing/extra/wrong. To deal with such scenario, maybe we would have to implement something like:
"""
pre-validation: validate user-provided pdict (validate only the keys! Disregarding values and data type)
initialize the MSPileupObj with the user-provided data
post-validation: validate self.data (expected object schema, but now validating values and value type).
"""

what do you think? Are we missing anything in here?

vkuznet replied:
I would like to avoid over-complicated steps. It should be one validation call, otherwise we'll need to maintain if/else logic in HTTP service APIs. I already thought about it, and if you checked my PR https://github.com/dmwm/WMCore/pull/11443 and look at 
- HTTP post request https://github.com/dmwm/WMCore/pull/11443/files#diff-5741164102ebe9ce2987f9e018708d8a39c8d3b8bc70701fd3446898e9189342R376
- and, `MSPileup` create API, https://github.com/dmwm/WMCore/pull/11443/files#diff-559a1ede5befe7a1ba862bb3a6cbf18fe85dcb9844fadb719f37da82db4e03d9R121
you'll see that it will call only `createPileup` data layer API, which in this PR does the following :
- create MSPileupObj (https://github.com/dmwm/WMCore/pull/11438/files#diff-df26107439b08bc216635653813392ddebb3dc94f0514c164a15d5115d790484R93) which by itself
- calls `validate` API, https://github.com/dmwm/WMCore/pull/11438/files#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR63

So, the HTTP post request will call create, the create will creates MSPileupObj which by itself will validate data. Therefore, if validate performs all the steps we do not need more complex logic. I may be wrong, but I think it is sufficient, and I do not think we need to over complicate existing data layer. But, if we'll find it later that it can be useful we can add such logic and split validate into multiple steps.

amaltaro replied:
In that case, your latest commit L93-L98 is not really needed, given that `pdict` in this validate method will always be equal to schema.
I would suggest to add a "NOTE" or "TODO" comment in the beginning of this method, saying that not all of the original user-provided payload is validated. Just in case we have problems in the near future and we need to improve it.

vkuznet replied:
I think the doc string of `validate` clearly describe the two cases, please see https://github.com/dmwm/WMCore/pull/11438/files#diff-21c86cd53529a8674a70ffc7953fbbfc5d8ff98acdd7bc6962d257bd55e673dcR82 I doubt we need anymore NOTEs or TODOs here, but we do need to get better understanding how we'll create and use objects. So far it is far from clear for me, and therefore it is better to keep the logic as is and rely on docstring which specifically describes the behavior.

amaltaro replied:
Okay, that works as well.

amaltaro reviewed (changes requested):
There you go.

vkuznet commented:
Alan, regarding your comment https://github.com/dmwm/WMCore/pull/11438#discussion_r1086757097, the unit test I provided clearly shows that both actual error is reported in a log and appropriate exception, e.g.
```
2023-01-25 11:04:02,808:ERROR:MSPileupObj: pileupName value pileupName does not match dataset pattern
2023-01-25 11:04:02,808:WARNING:MSPileupObj_t: expected exception MSPileup input is invalid
2023-01-25 11:04:02,808:ERROR:MSPileupObj: replicationGrouping value bla is neither of ['DATASET', 'ALL']
2023-01-25 11:04:02,808:WARNING:MSPileupObj_t: expected exception MSPileup input is invalid
.2023-01-25 11:04:02,809:ERROR:MSPileupObj: pileupName value pileupName does not match dataset pattern
2023-01-25 11:04:02,809:WARNING:MSPileupObj_t: expected exception MSPileup input is invalid
```
As you can see the `ERROR` message comes from `validate` function, while `WARNING` comes from unit test which catch the exception and show it. If you want actual exception to carry the error then the `validate` method should not return `True or False` but rather through exception. Please tell me which mode of operation do you want to have:
- `validate` will through the exception (new behavior)
- `validate` will return true or false but put actual error in logger error (current behavior)
Once I'll hear what do you prefer I can adjust code accordingly.

amaltaro commented:
> Please tell me which mode of operation do you want to have:
> *    validate will through the exception (new behavior)
> *    validate will return true or false but put actual error in logger error (current behavior)
>    Once I'll hear what do you prefer I can adjust code accordingly.

Hmm, I think we will need the exception to be raised (with the relevant error message), otherwise we cannot propagate what the actual error was back to the user (when interacting through a REST endpoint). In terms of the daemon service, boolean would be good enough because we already log it. Does it make sense to you?

vkuznet commented:
ok, I modified return of validate method to be `boolean, string` tuple where first value is a status and second is associated message. And, now, the unit test report the following:
```
2023-01-25 13:02:23,120:ERROR:MSPileupObj: pileupName value pileupName does not match dataset pattern
2023-01-25 13:02:23,120:WARNING:MSPileupObj_t: expected exception MSPileup input is invalid, pileupName value pileupName does not match dataset pattern
```
Please let me know if it accepted from your side, see https://github.com/dmwm/WMCore/pull/11438/commits/90d922ae118a8063bf062c7acb644d79ff0eef08

amaltaro commented:
@vkuznet your changes look good to me. I would just make a minor change, initialize msg (msg = "") at the very beginning of that method and remove the "" from the very last return statement

vkuznet commented:
ok, I made this change and squashed commits. Feel free to have a final look.

amaltaro reviewed (approved):
I think we had enough of reviews of this PR and it looks ready to go in. If we spot any enhancements or minor changes, we can always provide it within one of the future MSPileup PRs. If we spot major changes, then we should go with a new issue.
Thanks Valentin.
